### PR TITLE
Fix symbolic magic method lowerings

### LINKED ISSUE #127789: Inductor lowerings for magic methods are suspicious
### 🐛 Describe the bug

Steps to reproduce:
1. Patch in https://github.com/pytorch/pytorch/pull/126905 (landing soon)
2. Edit torch/_inductor/graph.py

```
            elif is_magic_method(n.target):
                # TODO: this is sus, it probably should be handled in the
                # lowerings themselves similarly to sym_size/sym-stride
                debug("is_magic_method")
                if isinstance(
                    n.meta["val"], (torch.SymInt, torch.SymFloat, torch.SymBool)
                ):
                    result = n.meta["val"].node.expr
                else: 
                    result = super().run_node(n)
```

remove the node.expr direct run check.

Tests will fail, because the lowerings will be doing insane things. Why are there insane lowerings in the table? I'm not sure...

### Versions

main

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184327

Normalize symbolic magic-method lowering inputs so Inductor no longer needs a graph-level shortcut for SymInt/SymFloat/SymBool results. Add dynamic-shape coverage for representative symbolic scalar magic methods.

Fixes #127789

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -8140,8 +8141,36 @@ def sym_numel(a):
     return a.get_numel()
 
 
+def _is_symbolic_magic_arg(x: Any) -> bool:
```
Comment (Skylion007) [inline torch/_inductor/lowering.py]: The return type here is typing_extensions.TypeIs of the two instance types

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -8140,8 +8141,36 @@ def sym_numel(a):
...
+
+def _unwrap_symbolic_magic_arg(x: Any) -> Any:
+    if isinstance(x, SymTypes):
+        return x.node.expr
+    if isinstance(x, (int, float, bool)):
+        return sympy.sympify(x)
+    return x
+
+
+def _make_magic_method_lowering(func: Callable[..., Any]) -> Callable[..., Any]:
```
Comment (Skylion007) [inline torch/_inductor/lowering.py]: The output of the Callable at least looks to be preserved? Might as well mark it as such with a _R TypeVar?

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -8140,8 +8141,36 @@ def sym_numel(a):
...
+    @functools.wraps(func)
+    def wrapped(*args: Any, **kwargs: Any) -> Any:
```
Comment (Skylion007) [inline torch/_inductor/lowering.py]: Should be _P.args and _P.kwargs

**Comment by jansel** [comment]
@claude review these changes

**Review by eellison — COMMENTED** [review]
not sure if cleaner or not, asking review of @ezyang

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (jansel) [reply torch/_inductor/lowering.py]: My agent says this was addressed by adding the _R TypeVar so _make_magic_method_lowering preserves the wrapped callable return type.

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/lowering.py]: My agent says this was addressed by changing _is_symbolic_magic_arg to return TypeIs[_SymbolicMagicArg] in the updated lowering helper.

File: torch/_inductor/lowering.py (hunk 3 above)
Comment (jansel) [reply torch/_inductor/lowering.py]: My agent says this was addressed by annotating the wrapper as *args: _P.args and **kwargs: _P.kwargs.

**Comment by ezyang** [comment]
@claude Find the historical commit that added the magic method workaround and explain the context in which it existed and why this was the easiest thing to do. And then explain what's different now.

**Comment by ezyang** [comment]
@claude Where is this newly added code modeled from?

```

def _make_magic_method_lowering(func: Callable[_P, _R]) -> Callable[_P, _R]:
    @functools.wraps(func)
    def wrapped(*args: _P.args, **kwargs: _P.kwargs) -> _R:
        node = V.graph.current_node
        meta_val = node.meta.get("val") if node is not None else None
        if isinstance(meta_val, SymTypes):
            return cast(_R, meta_val.node.expr)
        has_symbolic_value = any(
            _is_symbolic_magic_arg(x) for x in itertools.chain(args, kwargs.values())
        )
        if has_symbolic_value:
            normalized_args = tuple(_unwrap_symbolic_magic_arg(x) for x in args)
            normalized_kwargs = {
                k: _unwrap_symbolic_magic_arg(v) for k, v in kwargs.items()
            }
            return cast(Callable[..., _R], func)(*normalized_args, **normalized_kwargs)
        return func(*args, **kwargs)

    return wrapped

```

**Comment by ezyang** [comment]
@claude It just seems like there sure is a lot of boilerplate, just for a magic method where there are not that many variations

**Comment by jansel** [comment]
My agent says I simplified the magic-method lowering code in the latest push by removing the TypeIs/_R wrapper scaffolding and registering a small inline lowering with one unwrap helper. Local validation passed with the targeted CPU regression, py_compile, lintrunner -a, and git diff --check.

**Comment by jansel** [comment]
@claude review these changes

**Review by ezyang — APPROVED** [review]
The code here is still kind of sloppy but I'm not really sure what a better structure is

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main after the prior merge job timed out and DrCI pointed at stale trunk aarch64 failures plus a cancelled llm-retrieval job. I resubmitted the same code via ghstack; local validation passed with py_compile, lintrunner -a, git diff --check, and the targeted CPU regression using a stubbed torchvision import because this checkout's torchvision package has a CUDA major-version mismatch.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR onto current viable/strict after DrCI reported broken-trunk failures and docs failures from the recent OSDC direct-upload workflow where the docs jobs built successfully but failed at aws s3 sync with aws missing. The Inductor diff is unchanged; local source validation passed with py_compile, git diff --check, and lintrunner on the touched files with PYREFLY skipped because the local build/type state is mismatched with viable/strict's named-tensor state.

**Comment by jansel** [comment]
My agent says the current DrCI failures were Windows setup failures during requirements installation before any PyTorch build started. I rebased and resubmitted the unchanged Inductor fix onto current main to pick up the latest CI setup; local validation passed with py_compile, git diff --check, the targeted CPU regression using a stubbed torchvision import, and lintrunner -a.

**Review by mlazos — APPROVED** [review]

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the unchanged Inductor fix onto current origin/viable/strict after mergebot failed twice on the broken-trunk ROCm build. Local validation passed with py_compile, git diff --check HEAD~ HEAD, the targeted CPU regression using a stubbed torchvision import, and lintrunner -a.

**Comment by jansel** [comment]
Agent, the following is a test that should reproduce the reason for the revert.
1) Verify the test fails in the current version of the PR.  State clearly in your response if you were able to reproduce the failure
2) Add the test to PyTorch
3) Rebase the PR past the revert commit
4) Fix the test and update the PR

---


## How to create a unit test covering this case
The gap: this was a **codegen miscompile** (int64 symbolic scalar downcast to int32 via `index_expr`) that only surfaced at AOTI *runtime*, on GPU, hours into a cogwheel — far too late and too coarse. Two layers of coverage are warranted; **Level 1 is the one that would have caught it at diff-time.**
### Level 1 — PyTorch Inductor unit test (precise, cheap, the real gate)
Add/extend a dynamic-shapes Inductor test that exercises a **symbolic `int64` scalar that exceeds `INT32_MAX` flowing into a small-tensor kernel** (so the kernel's `index_dtype` is int32), and assert compiled == eager.
- File: `fbcode/caffe2/test/inductor/test_torchinductor_dynamic_shapes.py` (or alongside `test_op_dtype_propagation`, which PR #182872 already extended to `dynamic=True` — forward-porting that test is the canonical fix-coverage).
- Run on **GPU** with `dynamic=True` (the bug only appears when the value is an unbacked/symbolic int and the kernel picks int32 indexing).
```python
import torch
from torch._inductor.test_case import TestCase, run_tests
from torch.testing._internal.common_utils import (
    instantiate_parametrized_tests,
    parametrize,
)
from torch.testing._internal.inductor_utils import GPU_TYPE, requires_gpu
class TestSymIntIndexExprDtype(TestCase):
    @requires_gpu()
    @parametrize("dynamic", [True, False])
    def test_int64_symint_not_downcast_to_int32_index_expr(self, dynamic):
        """Regression for the AOTInductor illegal-memory-access:
        D107445917 (#184327, symbolic magic-method lowering) routed an int64
        symbolic scalar through index_expr, which used the kernel's int32
        index_dtype on small tensors -> overflow -> OOB index. Fixed by
        D107537584 (#182872)."""
        def fn(x, hi):
            # hi.item() is a data-dependent (unbacked) SymInt, int64, > INT32_MAX.
            h = hi.item()
            torch._check_is_size(h)
            torch._check(h > 2**31)
            # small output tensor => kernel index_dtype is int32; h flows via
            # index_expr. With the bug, h is downcast to int32 and overflows.
            return torch.randint(0, h, (x.shape[0],), device=x.device, dtype=torch.int64)
        x = torch.zeros(4, device=GPU_TYPE)
        hi = torch.tensor(2**31 + 1234, device=GPU_TYPE, dtype=torch.int64)
        cfn = torch.compile(fn, dynamic=dynamic, fullgraph=True)
        got = cfn(x, hi)
        # With the int32 overflow, the high bound becomes garbage/negative ->
        # values fall out of [0, hi) (or the kernel faults). Both must hold:
        self.assertTrue(bool((got >= 0).all()))
        self.assertTrue(bool((got < hi.item()).all()))
instantiate_parametrized_tests(TestSymIntIndexExprDtype)
if __name__ == "__main__":
    run_tests()

**Comment by jansel** [comment]
My agent says I rebased the PR past the internal revert and added the requested GPU regression coverage for an unbacked int64 randint high bound flowing through a small int32-indexed kernel. The new test checks both runtime bounds and generated Triton code (`ks1: i64`, `randint64`, and an int64 cast). Local validation passed with the focused GPU test, py_compile, git diff checks, and lintrunner -a; the direct dynamic-shapes file invocation still hits the known local torchvision CUDA 13.0 vs PyTorch CUDA 12.8 mismatch before the test body, so I also ran that CPU regression through a stubbed torchvision import.

**Comment by jansel** [comment]
@claude review these changes
